presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlock.java

```diff
@@ -299,89 +296,46 @@ public DictionaryBlock compact()
             return this;
         }
 
-        List<Block> compactBlocks = compactBlocks(singletonList(this));
-        return (DictionaryBlock) compactBlocks.get(0);
-    }
-
-    public static List<Block> compactBlocks(List<Block> blocks)
-    {
-        verifyEligibleToCompact(blocks);
-
-        DictionaryBlock dictionaryBlock = (DictionaryBlock) blocks.get(0);
-        Block dictionary = dictionaryBlock.getDictionary();
-        Slice ids = dictionaryBlock.getIds();
-
-        int positionCount = dictionaryBlock.getPositionCount();
+        // determine which dictionary entries are referenced and build a reindex for them
         int dictionarySize = dictionary.getPositionCount();
-
         List<Integer> dictionaryPositionsToCopy = new ArrayList<>(dictionarySize);
```

Comment:
Sometimes a smaller limit can be reached with `Math.min(dictionarySize, positionCount)`